### PR TITLE
change(mempool): Evict transactions from the mempool using the ZIP-317 conventional fee

### LINKED ISSUE #5335: Evict transactions from the mempool based on the ZIP-317 fee rules
### Motivation

We can deploy the new mempool rules in ZIP-317 before wallets have updated to the new conventional fee, because dropping transactions is randomised.

This can be changed in Zebra at any time. It must not be behind the `getblocktemplate-rpcs` feature, because it is required as soon as possible to mitigate denial of service:
https://zips.z.cash/zip-0317#deployment

### Specification

The conventional fee part of:
https://zips.z.cash/zip-0317

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

We want to implement the ZIP-317 conventional fee calculation, and use it to evict transactions from the mempool.

This is a security fix independent of the getblocktemplate work.
(But we also use the conventional fee calculation when selecting transactions for new blocks.)

Close #5335.

### Specifications

https://zips.z.cash/zip-0317#fee-calculation
https://zips.z.cash/zip-0317#mempool-size-limiting

This change can be deployed immediately:
https://zips.z.cash/zip-0317#deployment

### Designs

Do the fee calculation in the specified order, using similar names for everything:
https://zips.z.cash/zip-0317#fee-calculation

Cache the result as a field on `UnminedTx`, because it only changes when the transaction data changes. This way, it is available during block construction and transaction relaying.

This is a standard rule, not a consensus rule, so variations in the implementation are acceptable.

## Solution

- Add a ZIP-317 conventional fee module
- Add ZIP-317 constants
- Calculate the ZIP-317 conventional fee
- Use the new conventional fee for mempool size limiting
-  Update existing tests

Related changes:
- Just return a usize from `zcash_serialized_size()`, removing the unused `Result<_, io::Error>`
  - This makes the new code a lot simpler

## Review

I'd like two people to review this code, including @dconnolly because she helped write the spec.

Here is the corresponding code in `zcash_primitives` (it doesn't have Orchard support yet):
https://github.com/zcash/librustzcash/blob/main/zcash_primitives/src/transaction/fees/zip317.rs#L117

Do we need any extra tests?

### Reviewer Checklist

  - [x] Will the PR name make sense to users?
    - [x] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [x] Are the PR labels correct?
  - [x] Does the code do what the ticket and PR says?
  - [ ] How do you know it works? Does it have tests?

## Follow Up Work

The changes to transaction relaying and block production depend on enough wallets updating to pay these fees:
https://zips.z.cash/zip-0317#deployment

So they have separate tickets:
- #5473 
- #5336

### CODE REVIEW

File: zebra-chain/src/transaction/unmined.rs [hunk 1]
```diff
@@ -50,6 +52,12 @@ use UnminedTxId::*;
 /// [ZIP-401]: https://zips.z.cash/zip-0401
 const MEMPOOL_TRANSACTION_COST_THRESHOLD: u64 = 4000;
 
+/// When a transaction pays a fee less than the conventional fee,
+/// this low fee penatly is added to its cost for mempool eviction.
```
Comment (daira) [inline zebra-chain/src/transaction/unmined.rs]:
```suggestion
/// this low fee penalty is added to its cost for mempool eviction.
```

File: zebra-chain/src/transaction/unmined.rs [hunk 2]
```diff
@@ -327,31 +341,34 @@ impl VerifiedUnminedTx {
...
+    /// Standard rule:
     ///
     /// > Each transaction also has an eviction weight, which is cost +
     /// > low_fee_penalty, where low_fee_penalty is 16000 if the transaction pays
-    /// > a fee less than the conventional fee, otherwise 0. The conventional fee
-    /// > is currently defined as 1000 zatoshis
+    /// > a fee less than the conventional fee, otherwise 0.
     ///
+    /// > zcashd and zebrad limit the size of the mempool as described in [ZIP-401].
+    /// > This specifies a that is added to the "eviction weight" if the transaction
```
Comment (arya2) [inline zebra-chain/src/transaction/unmined.rs]:
```suggestion
    /// > This specifies a low fee penalty that is added to the "eviction weight" if the transaction
```

File: zebra-chain/src/transaction/unmined/zip317.rs [hunk 3]
```diff
@@ -28,5 +31,59 @@ const P2PKH_STANDARD_OUTPUT_SIZE: usize = 34;
...
+
+    let n_join_split = transaction.joinsplit_count();
+    let n_spends_sapling = transaction.sapling_spends_per_anchor().count();
+    let n_outputs_sapling = transaction.sapling_outputs().count();
+    let n_actions_orchard = transaction.orchard_actions().count();
+
+    let tx_in_logical_actions = div_ceil(tx_in_total_size, P2PKH_STANDARD_INPUT_SIZE);
+    let tx_out_logical_actions = div_ceil(tx_out_total_size, P2PKH_STANDARD_OUTPUT_SIZE);
+
+    let logical_actions = max(tx_in_logical_actions, tx_out_logical_actions)
```
Comment (arya2) [inline zebra-chain/src/transaction/unmined/zip317.rs:62]: Optional: move logical actions calculation to an `impl Transaction { fn count_logical_actions(&self) }` next to this fn

**Review by arya2 — COMMENTED** [review]
Code looks clean and correct, just needs minor updates to a couple doc comments.

I like that `zcash_serialized_size` was updated to return a `usize`.

File: zebra-chain/src/transaction/unmined/zip317.rs (hunk 3 above)
Comment (teor2345) [reply zebra-chain/src/transaction/unmined/zip317.rs:62]: I'd prefer to just have `conventional_fee` as a field, there's no other uses for logical actions yet, and the serialization of transaction data can be slightly expensive due to cryptographic operations when serializing types.

File: zebra-rpc/src/methods/tests/vectors.rs [hunk 4]
```diff
@@ -288,6 +289,7 @@ async fn rpc_getrawtransaction() {
                         id: UnminedTxId::Legacy(tx.hash()),
                         transaction: tx.clone(),
                         size: 0,
+                        conventional_fee: Amount::zero(),
```
Comment (arya2) [inline zebra-rpc/src/methods/tests/vectors.rs:292]: I think existing tests are sufficient but I'll leaving this here to give @dconnolly a chance to disagree.

**Review by arya2 — APPROVED** [review]
Thank you!

**Pull request merged**